### ISSUE TITLE
Async all the things

### ISSUE THREAD

**Issue by thecodejunkie** [NancyFx/Nancy#2576]
### Description
In `2.0-barneyrubble` and the upcoming `2.0-clinteastwood` released we've made changes on the async surface area we provide over some of our API, such as the [improved route declaration syntax](https://github.com/NancyFx/Nancy/pull/2441), but there are still places that needs to be updated to give us a complete async end-to-end experience.

## TODO (work in progress)
- [ ] [IResponseProcessor](https://github.com/NancyFx/Nancy/blob/master/src/Nancy/Responses/Negotiation/IResponseProcessor.cs) - For content negotiation

**Comment by khellang** [comment]
I suggest we start with the body delegate and then the async-ness will propagate from there. That's the bottom of the rabbit hole :stuck_out_tongue_closed_eyes:

**Comment by thecodejunkie** [comment]
@khellang Yeah that make sense. However I'd like for us to device a way to partition it so it does not have to be one huge PR with 5000 changes files and 50000 changes. Thoughts?

**Comment by khellang** [comment]
Yeah, I agree. There *might* be a way, but I think it involves the code not building for a while. Because async/await is so viral, if you go with the "bottom up"-approach, things will break and you'd have to convert everything at once. Maybe it's better to go top to bottom?

**Comment by grumpydev** [comment]
The body delegate is what I meant by adding "Response".. that whole class needs looking at, it's pretty wonky :)

**Comment by thecodejunkie** [comment]
@khellang perhaps we can artificially create boundaries by wrapping the delegate at some intersection(s) in the code?

**Comment by thecodejunkie** [comment]
@grumpydev what do you think we should change? It's a pretty light weight object right now

**Comment by grumpydev** [comment]
The way it works with the properties is kinda odd imo, but even if we don't change that, it should be an async delegate.

**Comment by thecodejunkie** [comment]
@grumpydev not saying you're wrong. I'm genuinely interested in exploring alternative designed for it. Make it immutable? Any other API changes? Strongly typed response headers etc. Perhaps spawn of a separate issue for the soul purpose of discussing / pseudo coding a new approach to `Response` ?

**Comment by grumpydev** [comment]
I think we already did it a while back, but it was a very invasive change, it's the one "big thing" that needs asyncing though for sure.

**Comment by thecodejunkie** [comment]
@grumpydev I think we should be able to change it to `Action<CancellationToken, Stream>` and wrap it in a sync delegate further up the stage where it branches out. Then we can one-by-one start removing the wrapper. Or have an `AsSync()` extension on it. So in an `IViewEngine` we could do `.AsSync()` to wrap it and it would still compile and function. Then we slowly start peeling away the use of the extension method. That should, in theory, mean we could attack it in sections without 50000 changes or a broken master for a while ?

**Comment by ravensorb** [comment]
Question, are there plans for supporting async for things like ApplicationStartup and RequestStartup?

**Comment by thecodejunkie** [comment]
@ravensorb how would that look / work though?

**Comment by ravensorb** [comment]
Fair question :) As more and more frameworks and supporting libraries are moving over to supporting (or completely using) Async, I am seeing a need to be more friendly to it in all areas.  For NancyFx, I could see either an "Task ApplicationStartupAsync(...)" method created that can be overriden or an event added for ApplicationStartUp that supports Async events.  Either way, the core framework would execute one of those in a way "nicely" blocks execution for the remainder of the pipelines until it completes.

Example of how it would look/be used:

```
public override async Task ApplicationStartup(ILifetimeScope container, IPipelines pipelines)
{
    await base.ApplicationStartupAsync(container, pipelines);

    await InitializeLoggingContainerAsync();
    await InitializeSessionContainerAsync();

    await RetrieveSettingsFromDbAsync();
}
```

**Comment by thecodejunkie** [comment]
Yeah, but we can't really continue execution of either of those methods, until they're done, because they're setting the stage for the rest of the execution 😄

**Comment by ravensorb** [comment]
Correct, so in the framework, there would need to be a "wait" for the completion.  Either using something like a await or ContinueWith
